lib/fileset/default.nix

```diff
@@ -278,4 +281,115 @@ If a directory does not recursively contain any file, it is omitted from the sto
         _unionMany
       ];
 
+  /*
+    Incrementally evaluate and trace a file set in a pretty way.
+    This function is only intended for debugging purposes.
+    The exact tracing format is unspecified and may change.
+
+    This function takes a final argument to return.
+    In comparison, [`traceVal`](#function-library-lib.fileset.traceVal) returns
+    the given file set argument.
+
+    This variant is useful for tracing file sets in the Nix repl.
+
+    Type:
+      trace :: { } -> FileSet -> Any -> Any
+
+    Example:
+      trace {} (unions [ ./Makefile ./src ./tests/run.sh ]) null
+      =>
+      trace: /home/user/src/myProject
+      trace: - Makefile (regular)
+      trace: - src (all files in directory)
+      trace: - tests
+      trace:   - run.sh (regular)
+      null
+  */
+  trace =
+    /*
+    An attribute set to configure how the file set gets printed.
+
+    Currently nothing can be configured, so this must always be an empty attribute set `{ }`.
+    */
+    config:
+    /*
+    The file set to trace.
+
+    This argument can also be a path,
+    which gets [implicitly coerced to a file set](#sec-fileset-path-coercion).
+    */
+    fileset:
+    let
+      # "fileset" would be a better name, but that would clash with the argument name,
+      # and we cannot change that because of https://github.com/nix-community/nixdoc/issues/76
+      actualFileset = _coerce "lib.fileset.trace: second argument" fileset;
+      # This is just one possible way to prepare a tree for tracing
+      minimalTree = _normaliseTreeMinimal actualFileset._internalBase actualFileset._internalTree;
+    in
+    if config != { } then
+      throw "lib.fileset.trace: The first argument must be { }. In the future this can allow configuration of the trace format."
```

Comment:
In this case, wouldn't it make more sense to go for the `foo`/`fooWith` pattern?
I find it hard to imagine a use case for parameters, whereas `filesets.trace myset` is going to annoy everyone, because they know `lib.trace`.

Reply:
Ohh that's a great suggestion. It annoyed me myself already when I forgot the `{}` argument.

It does explode the function space a bit, because you then have `trace`, `traceWith`, `traceVal` and `traceValWith`. I think that would be okay though.

I'll work towards this.

Reply:
You could refactor the function space with `concatMapAttrs` ;)